### PR TITLE
Normalized Riemannian optimizer

### PR DESCRIPTION
**Pull request by mkhona-nvidia**
Added an optimizer for normalized weights, i.e. weights whose columns or rows sum to 1. This is referred to as the "Oblique manifold" and this optimizer performs Riemannian descent on the oblique manifold (see An Introduction to Optimization on Smooth Manifolds by Nicolas Boumal at https://www.nicolasboumal.net/book/ for details)

### CODE REVIEW

File: emerging_optimizers/riemannian_optimizers/normalized_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+    References:
+        - An Introduction to Optimization on Smooth Manifolds (Nicolas Boumal)
+        - Jianlin Su: https://kexue.fm/archives/11196
+        - Raman et al.: https://arxiv.org/abs/1909.06463
+        - Franz Cesista: https://leloykun.github.io/ponder/steepest-descent-stiefel/#6-bonus-a-muon-like-optimizer-for-the-embedding-and-unembedding-layers
+    """
+
+    def __init__(
+        self,
```
Comment (skyw) [inline emerging_optimizers/riemannian_optimizers/normalized_optimizer.py:44]: missing type annotation.

File: emerging_optimizers/riemannian_optimizers/normalized_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+        params,
+        lr=1e-3,
+        momentum=0.9,
+        weight_decay=0.0,
+        mode="col",
+        eps=1e-8,
+    ):
+        """
+        Args:
```
Comment (skyw) [inline emerging_optimizers/riemannian_optimizers/normalized_optimizer.py]: Nit: No __init__ specific content, can move args to class docstring.

File: emerging_optimizers/riemannian_optimizers/normalized_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+                # Initialize momentum buffer if needed
+                state = self.state[param]
+                if "momentum_buffer" not in state:
+                    state["momentum_buffer"] = torch.zeros_like(param)
+
+                buf = state["momentum_buffer"]
+
+                # theory style momentum
+                buf.mul_(mom).add_(grad)
```
Comment (skyw) [inline emerging_optimizers/riemannian_optimizers/normalized_optimizer.py]: `buf = torch.add(grad, buff, alpha=mom)` would be more efficient.

File: emerging_optimizers/riemannian_optimizers/normalized_optimizer.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+                # Apply Riemannian gradient update
+                _compute_riemannian_grad_and_update(param, buf, mode, lr, wd)
+
+                # Retraction back to the manifold
+                if mode == "col":
+                    norm = param.norm(dim=0, keepdim=True).clamp(min=eps)
```
Comment (skyw) [inline emerging_optimizers/riemannian_optimizers/normalized_optimizer.py]: Use `dim` or `oblique_dim` as argument, so they can be combined to the same line. bonus is it also supports negative index when they are used correctly.

File: emerging_optimizers/riemannian_optimizers/normalized_optimizer.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+class ObliqueAdam(Optimizer):
+    """Adam optimizer for row- or column-normalized 2D parameters on oblique manifolds.
+
+    This optimizer adapts an Adam-like algorithm to work on oblique manifolds, where
+    parameters are constrained to have unit-norm rows or columns. It combines
+    adaptive momentum estimation with Riemannian gradient computation and manifold retraction.
+    """
+
+    def __init__(
+        self,
```
Comment (skyw) [inline emerging_optimizers/riemannian_optimizers/normalized_optimizer.py:119]: Fix: add type hint.

File: emerging_optimizers/riemannian_optimizers/normalized_optimizer.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+                    bias_correction1 = 1.0
+                    bias_correction2 = 1.0
+
+                norm_grad = (exp_avg / bias_correction1) / (exp_avg_sq.sqrt() / bias_correction2 + eps)
+
+                # Apply Riemannian gradient update
+                _compute_riemannian_grad_and_update(param, norm_grad, mode, lr, wd)
+
+                # Retraction back to the manifold (compute norm after weight decay)
+                if mode == "col":
```
Comment (skyw) [inline emerging_optimizers/riemannian_optimizers/normalized_optimizer.py]: Same as the other one.

File: emerging_optimizers/riemannian_optimizers/normalized_optimizer.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+
+                if correct_bias:
+                    # step size correction for ADAM moments EMA
+                    bias_correction1 = 1.0 - betas[0] ** step
+                    bias_correction2 = 1.0 - betas[1] ** step
+                else:
+                    bias_correction1 = 1.0
+                    bias_correction2 = 1.0
+
+                norm_grad = (exp_avg / bias_correction1) / (exp_avg_sq.sqrt() / bias_correction2 + eps)
```
Comment (skyw) [inline emerging_optimizers/riemannian_optimizers/normalized_optimizer.py:211]: Move this into `if correct_bias`, 
`norm_grad = exp_avg / (exp_avg_sq.sqrt() / eps)` is much easier to read and potentially fast.

question: should it be (bias_correction2 + eps) or max? I guess the former is consistent with Adam although latter may be better numerically?

File: emerging_optimizers/riemannian_optimizers/normalized_optimizer.py [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+        # Column oblique: inner products over rows (dim=0), shape (1, p)
+        dim = 0
+    elif mode == "row":
+        # Row oblique: inner products over columns (dim=1), shape (n, 1)
+        dim = 1
+    else:
+        raise ValueError(f"Invalid mode '{mode}'; expected 'col' or 'row'")
+
+    inner = (param * grad_like).sum(dim=dim, keepdim=True)
+    riem_grad = grad_like - param * inner
```
Comment (skyw) [inline emerging_optimizers/riemannian_optimizers/normalized_optimizer.py]: Use addmm

File: emerging_optimizers/riemannian_optimizers/normalized_optimizer.py [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+
+    # Add decoupled weight decay
+    param.mul_(1 - lr * wd)
+
+    # Apply update in-place
+    param.add_(riem_grad, alpha=-lr)
```
Comment (skyw) [inline emerging_optimizers/riemannian_optimizers/normalized_optimizer.py:242]: Check if two lines can be combined to one function.

File: tests/test_normalized_optimizer.py [hunk 10]
```diff
@@ -0,0 +1,473 @@
...
+    def get_oblique_parameters(self):
+        """Return parameters that should use oblique optimization."""
+        return [self.fc1.weight, self.fc2.weight]
+
+    def get_regular_parameters(self):
+        """Return parameters that should use regular optimization."""
+        return [self.fc3.weight, self.fc3.bias]
+
+
+class NormalizedOptimizerConvergenceTest(BaseTestCase):
```
Comment (skyw) [inline tests/test_normalized_optimizer.py]: Move this to its own file and to L1 test script.

File: tests/test_normalized_optimizer.py [hunk 11]
```diff
@@ -0,0 +1,473 @@
...
+import torch.nn.functional as F
+from absl.testing import absltest, parameterized
+from torch.utils.data import DataLoader, TensorDataset
+
+from emerging_optimizers.riemannian_optimizers.normalized_optimizer import ObliqueAdam, ObliqueSGD
+
+
+# Base class for tests requiring seeding for determinism
+class BaseTestCase(parameterized.TestCase):
+    def setUp(self):
```
Comment (skyw) [inline tests/test_normalized_optimizer.py:32]: This base class is meaningless.

File: tests/test_normalized_optimizer.py [hunk 12]
```diff
@@ -0,0 +1,473 @@
...
+        """Set random seed before each test."""
+        # Set seed for PyTorch
+        torch.manual_seed(42)
+        # Set seed for CUDA if available
+        if torch.cuda.is_available():
+            torch.cuda.manual_seed_all(42)
+
+
+class NormalizedOptimizerFunctionalTest(BaseTestCase):
```
Comment (skyw) [inline tests/test_normalized_optimizer.py]: meaningless "functional" in test name. all tests here are functional

File: tests/test_normalized_optimizer.py [hunk 13]
```diff
@@ -0,0 +1,473 @@
...
+        ("col_mode", "col"),
+        ("row_mode", "row"),
+    )
+    def test_oblique_sgd_preserves_norms(self, mode):
+        """Test that ObliqueSGD preserves row or column norms after one optimization step."""
+        # Create a 4x6 matrix for testing
+        matrix_size = (4, 6)
+
+        # Initialize with random values then normalize
+        param = torch.nn.Parameter(torch.randn(matrix_size, dtype=torch.float32))
```
Comment (skyw) [inline tests/test_normalized_optimizer.py]: Question: Does it need to be Parameter? and use .data later? Seems to me tensor is enough.

File: tests/test_normalized_optimizer.py [hunk 14]
```diff
@@ -0,0 +1,473 @@
...
+            final_norms = param.data.norm(dim=0)
+        else:  # mode == "row"
+            final_norms = param.data.norm(dim=1)
+
+        # All norms should be approximately 1.0 (unit norm constraint)
+        expected_norms = torch.ones_like(final_norms)
+        torch.testing.assert_close(
+            final_norms,
+            expected_norms,
+            atol=1e-6,
```
Comment (skyw) [inline tests/test_normalized_optimizer.py]: atol should be 0 when comparing values that expected to be 1. same for all other tests.

File: tests/test_normalized_optimizer.py [hunk 15]
```diff
@@ -0,0 +1,473 @@
...
+            rtol=1e-6,
+        )
+
+    def test_oblique_sgd_zero_gradient(self):
+        """Test that ObliqueSGD handles zero gradients correctly."""
+        matrix_size = (2, 4)
+        param = torch.nn.Parameter(torch.randn(matrix_size, dtype=torch.float32))
+
+        # Normalize columns
+        param.data = param.data / param.data.norm(dim=0, keepdim=True).clamp(min=1e-8)
```
Comment (skyw) [inline tests/test_normalized_optimizer.py]: Does normalize() do the same?

**Review by skyw — CHANGES_REQUESTED** [review]
Haven't finished reviewing all the tests. Will do another path when requested changes are done.

File: emerging_optimizers/riemannian_optimizers/normalized_optimizer.py (hunk 7 above)
Comment (mkhona-nvidia) [reply emerging_optimizers/riemannian_optimizers/normalized_optimizer.py:211]: adam actually has both options, the max one is called "amsgrad": https://openreview.net/forum?id=ryQu7f-RZ

but in practice the first is always used

File: tests/test_normalized_optimizer.py (hunk 14 above)
Comment (mkhona-nvidia) [reply tests/test_normalized_optimizer.py]: I get tiny errors:
Greatest absolute difference: 2.9802322387695312e-08 at index (1, 0) (up to 0 allowed)

File: tests/test_normalized_optimizer.py (hunk 15 above)
Comment (mkhona-nvidia) [reply tests/test_normalized_optimizer.py]: https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.normalize.html

yep looks like it. will replace

File: tests/test_normalized_optimizer.py (hunk 11 above)
Comment (mkhona-nvidia) [reply tests/test_normalized_optimizer.py:32]: I use this to set seed and device

File: emerging_optimizers/riemannian_optimizers/normalized_optimizer.py (hunk 8 above)
Comment (mkhona-nvidia) [reply emerging_optimizers/riemannian_optimizers/normalized_optimizer.py]: this is a pointwise op, cannot use addmm

File: emerging_optimizers/riemannian_optimizers/normalized_optimizer.py (hunk 9 above)
Comment (mkhona-nvidia) [reply emerging_optimizers/riemannian_optimizers/normalized_optimizer.py:242]: these two are already inplace, i could try something like cmul: `torch.addcmul(param * (1 - lr * wd), riem_grad, torch.tensor(-lr), out=param)` but this would create an intermediate tensor `param * (1 - lr * wd)`

**Comment by mkhona-nvidia** [comment]
\ok to test dd2194d5ab3a71fa3daaf2c4d9c05efa11ac2840

File: tests/test_normalized_optimizer.py (hunk 11 above)
Comment (skyw) [reply tests/test_normalized_optimizer.py:32]: It doesn't have to be a base class. Each test case can have its own setUp. also note `setUp` is called before each test not the test class init

File: tests/test_normalized_optimizer.py (hunk 12 above)
Comment (skyw) [reply tests/test_normalized_optimizer.py]: Nothing changed why it is resolved?

File: tests/test_normalized_optimizer.py [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+    def test_oblique_adam_zero_gradient(self):
+        """Test that ObliqueAdam handles zero gradients correctly."""
+        matrix_size = (2, 3)
+        param = torch.randn(matrix_size, dtype=torch.float32, device=self.device)
+
+        # Normalize
+        torch.nn.functional.normalize(param, p=2.0, dim=1, eps=1e-8, out=param)
+        initial_param = param.clone()
+
+        param = torch.nn.Parameter(param)
```
Comment (skyw) [inline tests/test_normalized_optimizer.py]: Parameter left over

File: tests/test_normalized_optimizer.py [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+
+        # Normalize
+        param = param / param.norm(dim=0, keepdim=True).clamp(min=1e-8)
+
+        param = torch.nn.Parameter(param)
+        optimizer = ObliqueSGD([param], lr=0.05, momentum=0.8, dim=0)
+
+        # Perform multiple optimization steps
+        for step in range(10):
+            torch.manual_seed(step)  # Different gradient each step
```
Comment (skyw) [inline tests/test_normalized_optimizer.py]: rand would be different each loop without reset seed.

File: tests/normalized_optimizer_convergence_test.py [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+        self.fc3 = nn.Linear(hidden_size, num_classes, bias=True)  # Final layer with bias
+        self.dim = dim
+        # Initialize weights for oblique optimization
+        self._initialize_oblique_weights(dim)
+
+    def _initialize_oblique_weights(self, dim):
+        """Initialize weights to be normalized for oblique optimization."""
+        with torch.no_grad():
+            # Normalize  of oblique layers
+            self.fc1.weight.data = self.fc1.weight.data / self.fc1.weight.data.norm(dim=dim, keepdim=True).clamp(
```
Comment (skyw) [inline tests/normalized_optimizer_convergence_test.py]: here you do want inplace.
If there is no inplace init function does this, use `self.fc1.weight.data /= fc1.weight.data....`

File: tests/normalized_optimizer_convergence_test.py [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+        """Return parameters that should use oblique optimization."""
+        return [self.fc1.weight, self.fc2.weight]
+
+    def get_regular_parameters(self):
+        """Return parameters that should use regular optimization."""
+        return [self.fc3.weight, self.fc3.bias]
+
+
+# Base class for tests requiring seeding for determinism
+class BaseTestCase(parameterized.TestCase):
```
Comment (skyw) [inline tests/normalized_optimizer_convergence_test.py]: Remove meaningless inheritance.

File: emerging_optimizers/riemannian_optimizers/normalized_optimizer.py (hunk 9 above)
Comment (skyw) [reply emerging_optimizers/riemannian_optimizers/normalized_optimizer.py:242]: inplace doesn't make it any faster. fine to leave it is if nothing better come up. torch.tensor(-lr) need to handle device.

File: tests/normalized_optimizer_convergence_test.py (hunk 19 above)
Comment (mkhona-nvidia) [reply tests/normalized_optimizer_convergence_test.py]: resolved

**Comment by mkhona-nvidia** [comment]
> Couple of left overs need further fix, parameter for example.
> 
> I didn't check convergence test very carefully. Also L1 tests haven't been set up in CI, post a local run result in the chat before final approval.

Here are results of convergence test:

Running tests under Python 3.12.8: /Users/mkhona/miniconda3/envs/pytorch_env/bin/python
[ RUN      ] NormalizedOptimizerConvergenceTest.test_oblique_adam_convergence
[       OK ] NormalizedOptimizerConvergenceTest.test_oblique_adam_convergence
[ RUN      ] NormalizedOptimizerConvergenceTest.test_oblique_sgd_convergence
[       OK ] NormalizedOptimizerConvergenceTest.test_oblique_sgd_convergence
[ RUN      ] NormalizedOptimizerConvergenceTest.test_optimizer_modes_convergence_adam_col
Final accuracy: 89.2
[       OK ] NormalizedOptimizerConvergenceTest.test_optimizer_modes_convergence_adam_col
[ RUN      ] NormalizedOptimizerConvergenceTest.test_optimizer_modes_convergence_adam_row
Final accuracy: 80.4
[       OK ] NormalizedOptimizerConvergenceTest.test_optimizer_modes_convergence_adam_row
[ RUN      ] NormalizedOptimizerConvergenceTest.test_optimizer_modes_convergence_sgd_col
Final accuracy: 74.8
[       OK ] NormalizedOptimizerConvergenceTest.test_optimizer_modes_convergence_sgd_col
[ RUN      ] NormalizedOptimizerConvergenceTest.test_optimizer_modes_convergence_sgd_row
Final accuracy: 100.0
[       OK ] NormalizedOptimizerConvergenceTest.test_optimizer_modes_convergence_sgd_row

File: tests/normalized_optimizer_convergence_test.py (hunk 18 above)
Comment (mkhona-nvidia) [reply tests/normalized_optimizer_convergence_test.py]: done

File: tests/test_normalized_optimizer.py [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+    """Tests for ObliqueSGD and ObliqueAdam optimizers that preserve row/column norms."""
+
+    def setUp(self):
+        """Set random seed before each test."""
+        # Set seed for PyTorch
+        torch.manual_seed(1234)
+        # Set seed for CUDA if available
+        if torch.cuda.is_available():
+            torch.cuda.manual_seed_all(1234)
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (skyw) [inline tests/test_normalized_optimizer.py]: Don't auto select device, especially in test. If the test is accidently assigned to a machine without GPU, it can end up never being tested on GPU. Same for other test cases.

**Review by skyw — CHANGES_REQUESTED** [review]
Rerun test after addressing last a few comment, plz.

**Review by skyw — APPROVED** [review]
